docs/redesign-proposal.md

```diff
@@ -0,0 +1,93 @@
+# Navigator design
+
+Navigator as a system intends to provide a way to easily provision, monitor and
+manage complex systems on top of cluster schedulers, initially targetting
+Kubernetes and OpenShift.
+
+## Architecture
+
+All communication between components of the Navigator system should communicate
+through the `navigator-apiserver`. This allows for a consistent audit trail,
+recorded in a central place. Monitoring services such as `kube-state-metrics`
+can then be used to track changes to the contents of the API server to be
+consumed by something like Prometheus.
+
+## Pilots
+
+Databases run by Navigator are wrapped by helper processes called 'Pilots'.
+
+Pilots contain the operational knowledge needed to operate their respective
+databases with the assistance of the `navigator-apiserver`.
+
+Each Pilot runs as PID 1 within the main application container of the
+application it 'manages'. This allows us to closely and carefuly monitor the
+lifecycle events and signals that the subprocess receives. This will allow us
+to run Pilot in environments other than Kubernetes, despite the individual
+properties of how these systems work.
+
+Because the Pilot runs before the actual application has started, it is able to
+perform any preflight or configuration checks at an appropriate time. This may
+include waiting for other Pilots to start, or determining a leader (with the
+help of the `navigator-apiserver`) that should begin the bootstrapping of a new
+cluster for that application. This also applies at exit time. The Pilot is able
+to trap signals sent to the container and ensure the underlying application
+exits cleanly, and if it doesn't, report this diagnostic information back to
+the `navigator-apiserver`.
+
+Pilots should watch the `navigator-apiserver` for changes to resources it
+requires, and use the apiserver to communicate with other instances of Pilot
+within the same application cluster (eg. within the same Elasticsearch
+cluster). This may include things like notifications that a scale down event
+has been requested, or even used to elect a 'leader' between the Pilots in the
+event one is required for the proper functioning of that application.
```

Comment:
* kubectl create -f elasticsearch1.yaml`
* `navigator-apiserver` stores resource
* `navigator-controller-manager` watches and responds to new ElasticsearchCluster resources
* `navigator-controller-manager` creates StatefulSet resources (+ anything else needed to fulfil the ElasticsearchCluster)
* ` kubernetes` creates pods & PVs for StatefulSet
* `Pod1` starts up
* `ElasticSearchPilot` connects to navigator-api and attempts to claim the leader lock

Now it's worth noting here that the lock that the Pilot acquires **IS NOT** the same as the Elasticsearch leadership election. It may be the case that Elasticsearch and Pilot leaders are in different pods. The Pilots will elect a leader *if they need to*. So in the Elasticsearch case, a leader is chosen to be the one instance of Pilot that sends cluster control events (such as draining a node of shards) eg. in response to a scale down. That leaders may in fact be different to whatever leader Elasticsearch has elected.

Again though, this is TBD and not solidified yet. & is an implementation detail of the Pilot itself (ie. this should not be Navigators concern directly). As far as Navigator is concerned, there's no such thing as a leader really. It provides a primitive that can be used by applications to *determine* a leader, but neither the `navigator-apiserver` or `navigator-controller-manager` should care whether or not a leader exists. Instead, each Pilot decides whether to take some action (eg. draining shards from a node) depending whether or not it is a leader.